sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -852,26 +852,142 @@ class Analyzer(
+   *       will be pulled out as the join conditions.
+   *    b. IN/NOT IN will be rewritten as semi/anti join, unresolved conditions in the Filter will
+   *       be pulled out as join conditions, value = selected column will also be used as join
+   *       condition.
+   *
+   * Note: CTEs are handled in CTESubstitution.
    */
   object ResolveSubquery extends Rule[LogicalPlan] with PredicateHelper {
 
-    private def hasSubquery(e: Expression): Boolean = {
-      e.find(_.isInstanceOf[SubqueryExpression]).isDefined
+    private def hasCorrelatedSubquery(e: Expression): Boolean = {
+      e.find(_.isInstanceOf[CorrelatedSubqueryExpression]).isDefined
     }
 
     private def hasSubquery(q: LogicalPlan): Boolean = {
-      q.expressions.exists(hasSubquery)
+      q.expressions.exists(_.find(_.isInstanceOf[SubqueryExpression]).isDefined)
+    }
+
+    /**
+     * Removes the conjunctive predicates of Filter that can't be resolved in this logical plan,
+     * returns the resolved new logical plan and removed predicates.
+     */
+    private def removeUnresolvedPredicates(q: LogicalPlan): (LogicalPlan, Option[Expression]) = {
+      val unresolvedConditions = ArrayBuffer[Expression]()
+      var hasOuterJoin = false
+      val removed = q transform {
+        case j: Join if j.joinType != Inner =>
+          hasOuterJoin = true
+          j
+        case f @ Filter(cond, child) if child.resolved && !f.resolved =>
+          if (hasOuterJoin) {
+            // The predicates inside a outer/semi join can't be pulled over join safely.
+            throw new AnalysisException("accessing columns of outer query inside join is not" +
+              " supported")
+          } else {
+            val (resolved, unresolved) = splitConjunctivePredicates(cond).partition(_.resolved)
+            unresolvedConditions ++= unresolved
+            if (resolved.nonEmpty) {
+              Filter(resolved.reduceLeft(And), child)
+            } else {
+              child
+            }
+          }
+      }
+      if (unresolvedConditions.nonEmpty) {
+        // try to resolve new logical plan and remove unresolved predicated again
+        val (removedAgain, moreConditions) = removeUnresolvedPredicates(execute(removed))
+        (removedAgain, (unresolvedConditions ++ moreConditions.toSeq).reduceLeftOption(And))
+      } else {
+        if (!q.resolved) {
+          throw new AnalysisException(s"subquery can't be resolved: ${q.treeString}")
+        }
+        (q, None)
+      }
+    }
+
+    /**
+     * Returns a resolved subquery and predicate that will be used to rewrite the IN subquery as
+     * semi join (predicate will be used as join condition).
+     */
+    private def rewriteInSubquery(
+        value: Expression,
+        subquery: LogicalPlan): (LogicalPlan, Expression) = {
+      val (resolved, joinCondition) = removeUnresolvedPredicates(execute(subquery))
```

Comment:
Has anyone ever used a Correlated In Subquery? E.G:

    SELECT   *
    FROM      a
    WHERE   a.id IN (SELECT id FROM b WHERE b.value < a.value)